### ISSUE TITLE
consider sampling improvements

### ISSUE THREAD

**Issue by adamfk** [adamfk/bouncy-button#7]
# TLDR
We may occasionally log an extra pulse. It seems rather rare at this point unless you connect a 200+ kHz self generated signal.

I'm not overly concerned about it. It would be nice to fully understand or have another person help test/validate.

# More details
It's difficult to know exactly how the AVR architecture works regarding the edge counting peripheral and digital input synchronization/sampling. Some things in the datasheet seem to contradict others and also earlier tests.

This [simulation](https://falstad.com/circuit/circuitjs.html?ctz=CQAgjOCmC0CcIAYB0CDMAOVBWVYsHYAmdfANlwXxABYqsQsEGYwwAoAWRHTFPDO6lq-PoRCEkWNr2FEmYAbGIjxCauhqIt9JgDMAhgBsAzpESduvVRvRDr48ZOlZ6hQsMJZR6DZ75MEZ3pqdXEvGjs-LUCASRBsMSjqKyiAyQY2OJCNDGEsZVzo9Kk8JiwlcSIQUh9KsSZ6jIAZEGg7dCZoMBDuBvAQAxMzHTYAdxoU8Oyw-zGJ0SmrQsDx-Jz0PIrlufKxQt3xWpWGHqiD1It8Dd7BYQ6HCSlxhJn4-Jvjl-uru4C5tw8tQBb3q-3cIJooRex2mL2mxA0gS453C5wQvkcUi4yQWfBx9kSTmRW2uB3xhKkACVWlD3l0wHt3hAxLwEGz2ZjOWkseJwS9gYUKRZgVEReiHkSQPg1HUpTLUIQ+o9nKgbmAKuqWRU+vQSowtAo+PdDVpGnrVfcEeAKla+mU5paqpbwsdnXxNeJqH9xh73PINdq5r7ar6qsdQ2JLeDXQ1wZbxYFSlpUDL7inhHaMuN-WIuu7A+M89bcykjnMi7bDoigzavbyPC7Mq07KyDaQAohinMmArOnZe+ZC-3Fc3hK3jkWXhWy4Wlu8Be8YUtrrDrkjLKI67YPHWhVweKInZEqkLqXB3e3WgyNOOHB12WzOY9wJJpfhYORCKR3+oP-4nIWtIlnw0JsAARq0xrKIqwioKg8CBAAHuIECsFQDKoSOO4gAAIgAOsYTQAIIACoAMIABJsMhXryOQ1QIBeeKeiAACiOEAOKsQROGsSRrFkSRADylLUVKGDgNQ8DoOEeBiNhZFCQAqgAcvxonPCOUT8o2EFwSy+Cqt+VDULgg7VECVTSh4YZiVcTBSYI8DvNhAAKMQqfhxhgQAlgALmJXSnAguawLGzFgBo7kqXheEAHYea5SkkYFDKwe+NIgeg8ngBoZFJQRZFNAA0nM1mynIspIVeoTlCA8CkDlkUgAAtvocUAK5GARPlxQADh1fkEQA9n1fk+cNcXlh+4CXm0Y6XoEw3iKqEAOWAIUMJ2D6wGQXieKaK3xLNBEcBRABehUFcYRWlctpn9Otm06Cg7K7aQ+06A4D2qolyVsMt3SPRMz2dmt-TRThAMGloySg8g4MQIpqnqWwQA) is my current understanding.

The problem still doesn't quite make sense to me. If my simulation/model is accurate, then there should be certain times when we sample the pin as low and the next CPU instruction, we sample the rising edge count as increased. This does happen if I self test with a frequency between 210 kHz (`gen a37`) and 381 kHz (`gen a20`). How often we see it drops off as frequency decreases.

![image](https://github.com/adamfk/bouncy-button/assets/274012/b6fd1f12-f8f2-4d2c-bfe3-9df347f2a034)

If it is a simple lining up issue, we should see it happen with a probability of roughly 1/40. However, when I test with frequencies slower than `gen a37`, it runs a lot without detecting any issue.

I'm also running a frequency sweep with my picoscope and an NPN (and external 1k pullup). It started at 100 kHz (the fastest it can) and has run for hours already. I modified code to not wait for stable or even detect initial change so it would go faster and hopefully remove any factors preventing reproducing the 1/40 issue. `gen a30` does reproduce the issue right away even with the modifications.

![image](https://github.com/adamfk/bouncy-button/assets/274012/8f830240-b266-4255-8782-16bd5de2e880)

# More
Interesting that `gen a38` gives a period of almost 2x constant logging speed of around 40 instruction cycles. Perhaps this issue is related to that.

Should try pattern generator that can vary spacing between pulses.
